File: src/main/kotlin/stats/StatsIdentity.kt
package com.tbread.stats

import java.security.MessageDigest

object StatsIdentity {
    const val IDENTITY_HASH_VERSION = "sha256:aion2-character:v1"

    fun characterIdentityHash(server: Int, nickname: String?): String? {
        if (server <= 0) return null
        val normalizedName = nickname?.trim()?.lowercase().orEmpty()
        if (normalizedName.isBlank()) return null
        return sha256("$IDENTITY_HASH_VERSION|$server|$normalizedName")
    }

    fun sha256(raw: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(raw.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}

File: dotnet/src/WaffleMeter.Stats/StatsIdentity.cs
using System.Security.Cryptography;
using System.Text;

namespace WaffleMeter.Stats;

/// <summary>
/// Verbatim port of Kotlin <c>stats.StatsIdentity</c>: a stable, anonymous per-character identity
/// hash = SHA-256 of "<c>version|server|lowercased-trimmed-nickname</c>", lowercase hex. Used to
/// de-duplicate uploads and tie consent to a character without sending the raw nickname.
/// </summary>
public static class StatsIdentity
{
    public const string IdentityHashVersion = "sha256:aion2-character:v1";

    public static string? CharacterIdentityHash(int server, string? nickname)
    {
        if (server <= 0)
        {
            return null;
        }

        string normalizedName = nickname?.Trim().ToLowerInvariant() ?? string.Empty;
        if (string.IsNullOrWhiteSpace(normalizedName))
        {
            return null;
        }

        return Sha256($"{IdentityHashVersion}|{server}|{normalizedName}");
    }

    public static string Sha256(string raw)
    {
        byte[] digest = SHA256.HashData(Encoding.UTF8.GetBytes(raw));
        return Convert.ToHexStringLower(digest);
    }
}
